### ISSUE TITLE
Deep packet inspection to classify V2Ray traffic

### ISSUE THREAD

**Issue by rickyzhang82** [v2ray/v2ray-core#2185]
In last 3 months, I'm verifying  [the patent claim](https://files.catbox.moe/vmzj04.pdf) by replicating it with code and data.

First, I spent two months to write a [TCP packet sorter](https://github.com/rickyzhang82/PacketSorter).  Then I collected home traffic for a month where I mixed v2ray traffic with other normal network traffic.

Last night, I stopped capturing packet data and started to build a classifier. I spent two hours to write a 1D Convolution Neutral Network to do binary classification on v2ray traffic and non-v2ray traffic. 

To my surprise, I can achieve classification accuracy rate up to 98.86%. In a nutshell, the patent claim is real. 

I shared my [classification model in Python notebook](https://github.com/rickyzhang82/V2Ray-Deep-Packet-Inspection/blob/master/V2Ray%20Deep%20Packet%20Inspection.ipynb) and findings in [my blog post](https://fr33land.net/2020/01/18/we-are-fucked/).

**Comment by twskipper** [comment]
interesting

**Comment by eycorsican** [comment]
What kind of V2Ray configurations were used in the data collection stage?

If a "tcp transport" was used, I am not surprised a DNN model can classify between non-HTTP/HTTPS data (data encrypted with VMess in your case) and HTTP/HTTPS data with a high score.

If a "ws+tls transport" was used, a DNN model could still learn to classify between different TLS fingerprints. (https://github.com/v2ray/v2ray-core/issues/2098)

DNN models are powerful, but if you are not using the right dataset, you are making a wrong assumption.

As the first step, maybe you could try to include more diverse data in your test set, e.g. data encrypted with Shadowsocks, replicate some VMess encrypted data and re-fill them with random bytes (keep the length), mark them as non-V2Ray data, and see if there is a performance drop.

**Comment by rickyzhang82** [comment]
@eycorsican 

I used vmess on TCP. I just released packet post processing repository. 

If you are interested in it, you can replicate my experiment on your own V2Ray configuration and your own traffic.

**Comment by eycorsican** [comment]
> If a "tcp transport" was used, I am not surprised a DNN model can classify between non-HTTP/HTTPS data (data encrypted with VMess in your case) and HTTP/HTTPS data with a high score.

I am not going to reproduce your work. If you can't eliminate the above possibility, your work is meaningless, IMO.

**Comment by rickyzhang82** [comment]
**Meaningless** is a strong word, kid.

I placed PacketSorter in my pfSense router over two months. The non-v2ray traffic includes OpenVPN, ssh, SMTP, IMAP, http, https, video streaming like Netflix, iTune and Youtube. It is a typical American family home network traffic.

There are naysayers and doers in our world. You are free to make your own choices.

**Comment by eycorsican** [comment]
Good, you gave more information about your dataset.

Nowadays many protocols are using TLS (including OpenVPN), header fields of the TLS protocol are salient features that DNN models would love to learn from, just looking at the version number field would achieve a high classification rate for TLS traffic, they're easy to learn compare to VMess's "random" bytes. So why not a DNN model just learn to distinguish these structured-data from "random" bytes? I believe almost all "normal" web traffic are using structured-protocols, they're all easy to learn by a DNN model.

But, **"random" bytes need not be VMess data**.

If you are not allow someone arguing your results and keep saying "you can replicate my experiment", why you're posting your results here.

**Comment by rickyzhang82** [comment]
Am I not allowing you to argue here? I think this is a privilege that being abused by the moderators here very often. I don't want to waste anyone's time here to trade fight. You start behave like a fifty cents. I wish you could speak with a proper manner like a decent human.

The vmess protocol encrypts the very first packet after TCP 3-way handshake. Within the first 16 non-empty TCP payload packet in the session, it stands out from encrypted protocol like SSH and TLS where clear text key exchange are easily identified.

The ROC curve in the experiment shows both type I error and type II error are low. That answers your question 

> But, "random" bytes need not be VMess data.

When you asked me to hand craft packets like the following, do you understand what is your purpose?

> replicate some VMess encrypted data and re-fill them with random bytes (keep the length), mark them as non-V2Ray data, and see if there is a performance drop.

Such network traffic doesn't really exist in the real world. Like you photoshop a monster with Emperor Xi's head and Winnie the Pooh's body, you ask the classifier: is this Emperor Xi or Winnie the Pooh? What is your point really?

Knowledge is important. However, those VMess, https, CNN or whatever dooms to be outdated within a decade. Learning to be a rational thinker and a decent man is far more important than any of this fashionable techie thingy.

Just my 2 cents.

**Comment by eycorsican** [comment]
> **If you can't eliminate the above possibility**, your work is meaningless, IMO.

It seems the word **meaningless** hits you hard, but I am using the right word and I have my precondition.

Well, I put my arguments quite clear, if you don't get them, that's fine.

V2Ray users will have their own judgements and make their own choices.

**Comment by GreatBigWhiteWorld** [comment]
Chill guys. Both of you are comrades fighting against the evil, which I can only admire without much contribution! The research is very interesting so let's stick to the technical and hope some others to chime in too.

**Comment by rickyzhang82** [comment]
Today, I did another round of tests. 

I spin up a Google Compute Engine (GCE) to deploy V2Ray server. It is completely different IP, port, UUID and OS from Amazon Lightsail. I used previously trained model by V2Ray server in AWS to classify traffic from V2Ray server in GCE.

The classification still can achieve (loss, acc) -- `[0.023288102071980347, 0.99675000000000002]`. The ROC curve still looks perfect.

Regarding vmess with TLS, I followed the [user guide](https://toutyrater.github.io/advanced/tls.html). But from the captured packets by Wireshark, I don't see clear text client hello, server hello or even key exchange like HTTPS. It seem the very first packet is encrypted already. Am I missing something? Or does vmess with TLS masquerade vmess protocol inside TLS?

**Comment by rickyzhang82** [comment]
I did another round configuration trials with TLS today. 

In vmess protocol, TLS with websocket `ws` or TLS with http/2 `http` shows a typical TLS clear text communication-- client hello, server hello but without server/client key exchange like HTTPS in our web browser. But TLS with `tcp` shows very similar encrypted packet pattern like vmess over TCP without TLS.

For sure, pure vmess over TCP can be easily classified now from my previous research. But it would be interesting to see if TLS with websocket and TLS with HTTP/2 can be masqueraded with other legitimate web browser HTTPS traffic. From my initial observation in Wireshark, the answer might be not. We will see.

**Comment by Cwek** [comment]
I amn't sure if it can use your training model to find the correct answer from a large number of TCP links without giving the correct conclusion (that is, the IP address and port required to deploy the service)? This may include many unfamiliar TCP links, and some may use data headers with similar patterns. Can it identify them? 

TLS does what it is supposed to do-it loads Vmess data streams into TLS data streams. This may also check whether this multi-layer load transport mode has the same or similar communication pattern to be identified.

As far as I know, the introduction of TLS 1.3 is based on user feedback, while older versions of V2ray  still use TLS 1.2. Whether the use of the TLS version of TLS 1.3 can be used as a pattern depends on the popularity of TLS 1.3. Some websites still introduce it as a test feature.

Due to the disappearance of the main developers, it may be difficult for the "night guard" developers to bring about feature changes to this protocol.

**Comment by Cwek** [comment]
V2ray v4.18.1 introduced TLS 1.3.  You can check V2ray of older version to check TLS 1.2.

**Comment by rickyzhang82** [comment]
@Cwek 
The configuration I used in my Python notebook is vmess over TCP.  The data set is collected over one month period in my home router. It represents a typical north American home network traffic. IP and port from both source and destination have been masked as zero value. So the trained model is immune from IP and port change.

The previously trained model by the data in AWS can classify the data in Google Compute Engine.

Machine learning is about data. That's why I release my infrastructure. Let anyone collect their own traffic in mainland China and train and evaluate the model on your own.

From my observation, vmess over TCP should **NOT** be recommended by default. It can be classified with high true positive rate and low false positive rate.

Regarding to the TLS, I'm reading V2Ray source code. It used Golang's  `crypto/tls` implementation and force to use TLS 1.3 through environment variable. I'm not sure if the model can classify vmess over TLS + websocket or vmess over TLS + http/s yet. But forcing everyone to use TLS 1.3 might not be a good idea while the majority TLS traffic is still in 1.2 version.

```
func init() {
	// opt-in TLS 1.3 for Go1.12
	// TODO: remove this line when Go1.13 is released.
	if !strings.Contains(os.Getenv("GODEBUG"), "tls13") {
		_ = os.Setenv("GODEBUG", os.Getenv("GODEBUG")+",tls13=1")
	}
}
```

I'm not concern with how to get back TLS 1.2. I'm curious if CNN can classify traffic with TLS combination.

**Comment by ailinwang** [comment]
> 
> 
> Good, you gave more information about your dataset.
> 
> Nowadays many protocols are using TLS (including OpenVPN), header fields of the TLS protocol are salient features that DNN models would love to learn from, just looking at the version number field would achieve a high classification rate for TLS traffic, they're easy to learn compare to VMess's "random" bytes. So why not a DNN model just learn to distinguish these structured-data from "random" bytes? I believe almost all "normal" web traffic are using structured-protocols, they're all easy to learn by a DNN model.
> 
> But, **"random" bytes need not be VMess data**.
> 
> If you are not allow someone arguing your results and keep saying "you can replicate my experiment", why you're posting your results here.

I would say it is indeed meaningless to launch a fruitless quarrel, what makes sense is that one formulate a well-posed question, such as : "Does the dataset really reflects a real family data transfer?" or "does this trained classifier work in another network?", a corresponding experiment will solve this problem.
It is therefore suggested to open a new board for discussion, any security flaw counts as it will **CERTAINLY** affect the efficacy of the anti-censorship software.
Machine learning is currently posing **MUCH** threat to cybersecurity.

**Comment by rickyzhang82** [comment]
@ailinwang 

I second your point.

As I pointed it out, the data I trained my classification model is a typical North American home network traffic. 

I released all data collection and processing tools and classification model source code. Because I expect **those live inside the mainland China to replicate the test with their own data**. Hopefully someone can tell me something with **empirical** data rather than calling other's work meaningless here.

Pretending to be a knowledgable guy is easier than getting your hands dirty to do something meaningful to our world.

**Comment by ailinwang** [comment]
> 
> 
> @ailinwang
> 
> I second your point.
> 
> As I pointed it out, the data I trained my classification model is a typical North American home network traffic.
> 
> I released all data collection and processing tools and classification model source code. Because I expect **those live inside the mainland China to replicate the test with their own data**. Hopefully someone can tell me something with **empirical** data rather than calling other's work meaningless here.
> 
> Pretending to be a knowledgable guy is easier than getting your hands dirty to do something meaningful to our world.

V2RAY is a rather complicated tool involving multiple protocols, different combination of protocols can generate different packets and thus result in different features.
Let me sum up your experiment: you conducted your experiment on ur own dataset collected in ur home in the US for 2 weeks. You have several clients that connect the Internet via a pfSense router which has a copy of V2RAY installed, the data was collected in the upstream port of your router. Then you train your data, I cannot find the detailed arch of your network in ur blog post, you predict with ur model trained and got a 98%~99% of accuracy. B.T.W, how did you partition your test set and training set? and how good is its generalization capability?
I think it is better to write a well formatted and rigorous paper to illustrate the experiment, including data collection procedure, your training procedure, and means of evaluation, your analysis, etc.
Besides, I would greatly appreciate it if you can provide a counter solution thru **PR or algorithm proposal** to lower its detection rate or to avoid any possible detection.

**Comment by rickyzhang82** [comment]
In the first experiment, I used pure VMESS over TCP settings.

The model is in [Python notebook](https://github.com/rickyzhang82/V2Ray-Deep-Packet-Inspection/blob/master/V2Ray%20Deep%20Packet%20Inspection.ipynb). You can see how I participate the test set and training set in data generator and how I collect data.

Also, Please read [my last blog post](https://fr33land.net/2020/02/06/we-are-fucked-indeed/). I used pre-trained classifier to test on unseen V2Ray server traffic where V2Ray server moved from AWS to Google Compute engine, different OS flavor and different UUID. It still achieve near perfect ROC curve. I'm not concern the model is over-fitting.

I'm not saying V2Ray should not be used now. But we should **NOT** use VMESS protocol over TCP. Big brother know you are using it now without decryption.

To counter classifier, I'm doing research on adversarial example to evade the classifier. So far I have tried two gradient based approaches, namely FGSM and DeepFool, to add noise in the first 16th non-empty TCP payload packet. The adversarial example generated by FGSM is easy to compute and effective while the one by DeepFool is not easy to compute due to the vanishing gradient in activation function of output layer. 

The hard part is to figure it out how to apply this adversarial noise in our communication settings, where we have domain specific constraints such as how to encode and decode the noise in two parties and how to apply noise effective in TCP payload only (You can't perturb the TCP and IP headers)

In any case, I will publish more research in my blog once I sorted out a feasible solution.

**Comment by rickyzhang82** [comment]
I published [new blog post on white-box adversarial example](https://fr33land.net/2020/02/12/white-box-adversarial-example/). There is a Python notebook that I wrote adversarial noise generation in Tensorflow 2.0.

Under three assumptions below, we could evade DNN classifier by adding adversarial noise. In other word, the V2Ray traffic with adversarial noise will be miss-classified as non-V2Ray traffic.

1. We own the classifier model.
1. We are oracles. Thus, we know the first 16th non-empty TCP payload traffic ahead.
1. We can apply adversarial noise without any restrictions.

My next step is to remove this assumption one-by-one.

**Comment by suhuaixing** [comment]
Excellent work indeed, you know why this is so important for us Chinese? 
Millions of Chinese are using v2ray right now(plus those who don't set it up by themselves but by sharing).
You did prove vmess only(TCP) could be easily identified, UDP is also true as our experience. This is not meaningless, at least you are warning us not to use v2ray+tcp only.
Indeed, this has already been proved being identifiable a long time ago as a common sense for us, so many people will say your work is meaningless. 

You know what, last year, a Beijing grad student applicated a patent about using DNN to identify SS or V2ray, which made a huge argument half a year ago. 99.9% of people were cursing these guys. Since this is almost our only way to bypass GFW. You may live in the States, you couldn't imagine how important a secure way to bypass GFW is to us.

Nowadays, many of us are using v2ray + ws + tls + caddy(or nginx). V2ray listens on non-80 and 443 port , say port 12345, then caddy or nginx listens on port 443, hosting a normal site on port 443 with normal domain, say v2ray.com, and a path parameter url like (/h2). If a client visits v2ray.com/h2, nginx will proxy the traffic to the real port 12345. If the password is not right, v2ray sends back a bad request, it doesn't mean anything, since this is not a trace which could be used as identification. With tls, the client's visit url won't include /h2 in the end, but only the domain, so the url is safe.


If you are instrested, you can test this config and I have two suggestions for you. If you can still identify this domain is using as a v2ray proxy server, not a normal site(say a static site). This could be huge. A dynamic site or download site or video site or even a real business site should be also tested to make sure if it is a "normal" site.
There's a tutorial and script[1] teaching how to set up a v2ray + ws + tls + caddy (or nginx) + (cdn) server.

As our experience told us, with cdn, a v2ray server will be pretty secure, since this is regarded as unblockable. Even GFW know you are bypassing GFW, it couldn't block a cloudflare cdn, it's about bussiness, millions of sites are using cloudflare cdn.

So, I'd suggest you test it without cdn. I haven't heard anyone's VPS being blocked by using this configuration, but technically, GFW can block this ip, not like with cdn. We don't know how secure it is indeed. Thanks.

The second point is, there's someone saying that h2 is not secure as ws. So this is a good point to test it. V2ray + h2 + tls + caddy(or nginx). The reason may be due to h2's traffic is not like h/1.1, it may be a weak point.

You know how we got this sense? Many of us are using different configurations, if someone's VPS is blocked, we'd know this is not secure. Hah, so in my mind, this is the secure level as below. GFW has a method called TCP blocking, so I think if there's only tcp vs udp, UDP is better. But, indeed, they're both bad, UDP could be blocked too, and many times, GFW will directly block the ip, which is even worse. As for obfuscation.. I'd say there's no way a real wechat video traffic would behave like v2ray with obfuscation, so it's pretty insure.
V2ray + ws + tls + caddy + cdn => 100% secure
V2ray + ws + tls + caddy => 95%
V2ray + h2 + tls + caddy => 90%
V2ray + udp => 20%
V2ray + tcp => 10%
V2ray + udp + obfuscation => 5%


[1] https://armazopu.github.io/v2ray+websocket+tls+caddy+cdn.html

**Comment by rickyzhang82** [comment]
You don't need to tell me why this is important. If this were not important, I would not spend my spare time in this. 

**Free speech** is the only cure to the sick country.

There are many configuration settings in V2Ray. It takes time to collect traffic data with one particular settings for machine learning purpose. I spent a month to collect network traffic with vmess over TCP setting in my first report.

I started to collect a new set of data with vmess + TLS configuration on Jan 29th. Today, I just found that there was a power outbreak on Feb 6th from my home server apcupsd daemon log. My pfSense router is running on a retired PC which doesn't connect to UPS. The power outbreak interrupted the data collection. Now I missed seven day long data collection.

There is some minor setback in this.

```
Feb 06 13:59:41 t20 apcupsd[1729]: Power failure.
Feb 06 13:59:43 t20 apcupsd[1729]: Power is back. UPS running on mains.
Feb 06 13:59:52 t20 apcupsd[1729]: Power failure.
Feb 06 13:59:54 t20 apcupsd[1729]: Power is back. UPS running on mains.
Feb 06 14:00:06 t20 apcupsd[1729]: Power failure.
Feb 06 14:00:07 t20 apcupsd[1729]: Power is back. UPS running on mains.
Feb 06 14:00:08 t20 apcupsd[1729]: Power failure.
Feb 06 14:00:10 t20 apcupsd[1729]: Power is back. UPS running on mains.
Feb 06 14:00:24 t20 apcupsd[1729]: Power failure.
Feb 06 14:00:26 t20 apcupsd[1729]: Power is back. UPS running on mains.
Feb 06 14:00:46 t20 apcupsd[1729]: Power failure.
Feb 06 14:00:48 t20 apcupsd[1729]: Power is back. UPS running on mains.
```

There are tons of smart people in mainland China. Can anyone replicate this experiment on your own V2Ray settings and traffic data? I knew I repeated myself. But God helps those who help themselves.

TBH, I don't need V2Ray at all. I have no ideas how you setup V2Ray. Neither do I have the similar traffic pattern like yours. We need collective efforts to beat big brother. They have the trove of data to train their model. But I can't fight on this alone.

If this were well-known that vmess over TCP was insecure before my releasing report, why it is still recommended settings in v2ray.com user guide? Why not deprecate the settings out of the box in V2Ray?

One more note, I'm not sure how you can get the percentage of secure on the V2Ray configuration. We need an experiment to verify the claim. This is science not religion. The vmess over TCP is **not** secure at all. This has been proven. IMO, security in software is a binary thing: either secure or insecure. There is no gray area.

**Comment by suhuaixing** [comment]
really thanks your intrests, but the reality is, this is complicated, many things are complicated. Some are about software, some about business, and some about politics. There's some truth.

v2ray.com as the offical doc, which couldn't be too bised. As a technique, it mostly only offer a choice, not how to efficiently prevent blocking. Even there is, it's pobrobally coming from community guide, and most of them are coming from users themselves.

There's too many unknown variables which could cause distinction but being ignored by people, say, V2ray version, OpenSSL version and too many, maybe one hundred more. Even we use the same os and software, but everyone's so-called normal usage is drastically different. Say, some software from Tencent or 360 safe or whatever, they can search user's disk and monitor network traffic to help checking if you are bypassing gfw. One main flaw about DNN is that, it couldn't tell us why it's so certain about it does being v2ray. A "normal" daily usage will include too many uncertainty, which is not much helpful for developing.

Some places are more like being blocked, I couldn't name it here, but as we know, it's more about place but what you are using. Sometimes, even we just use tcp or udp, it can hold a long time. Sometimes, if there's some big social events, gfw could be much more active than other days, say right now, due to coronovius, it's pretty active. we could just guess.
Thus, we simply couldn't doing research like a academic problem. Anyconnect is a pretty bad vpn tech, we can very easily tell which is using it, but gfw won't block it since many international companies are using it. So, we are not judging by how distinguishable it is but by how many and who are using it. If v2ray have millions of users, gfw simply won't block it too often. These common sense are 
gossiped in telegram or in private conversitions, it's more like a battle not research, British people are not gonna discuss encoding with Germans in WWII. If someone is doing secret research in public, it will be used by the opponents, only gfw will use these techniques to identify traffic. Since DNN couldn't tell the reason and which part is the key problem. You may think it's like a GAN. But GFW is not working like this, first it's probaly not using DNN but others like phishing, active trying, decoding, even with software helping. Sometimes, it will block a lot ips together not a single one. And we are not doing gene reasearch. Since gfw has tons of paid employees too, which is constantly changing and upgrading.
I'd suggest we don't do this like a normal research, if you're doing papar, it should be more specific. And bypassing gfw is not a scientific research but more like a social research, experiences are more imporant than analyzing. If you get a conclusion which is inconsistent with exps, it'll cause more horror than helping. And you are more like helping gfw indeed, maybe they don't use DNN, but after seeing your paper, they will develop one like this way.

**Comment by rickyzhang82** [comment]
I'm not interested in publishing paper. I'm more like a garage scientist. My motivation is to defeat censorship in all forms. Even in a free society like America, we still have censorship like political correctness. But I use my vote in America rather than the scientific web browsing technique you use in China to defeat it.

I understood some of your points. But one thing I want to point out is that we should make our research public. In the encryption field, the most secure way to design your encryption algorithm is to publish your algorithm. Only in this way, any other eyeballs can help find the flaws in its principles and the bugs in its implementation. Keeping your encryption in the safe doesn't make your encryption secure. I think Nazi and Jap learned their lessons from WWII. The modern asymmetric encryption algorithm is based on the hard math problem -- a large number factorization. Its encryption principle and implementation are in the public domain. We use the well-known NPC problem to defend our encryption. But we no longer keep the algorithm under our mattress as Nazi or Jap did.

We don' know if big brother uses DNN or not to classify network traffic now. But these techniques are feasible in reality. If the Chinese scholars applied for the patent, it is more likely that they are going to commercialize this technique in the near future. That is how the software company or patent troll company protects its IP in the West.

To accelerate the research on finding a way to defeat it, it is better that we open our research and let other smart minds join the fight.

**Comment by suhuaixing** [comment]
Let's save time. if you are interested, I'd suggest you do some more research, and don't get a conclution too quickly. You know this is about millions of people daily life and business. You can search gfw[1] in github, there's more than 1300+ projects you can learn about. Every Chinese programers know gfw and have already done plenty of awesome projects. They wrote code and made it public, which requires tons of their own time, they are great. As a user, we shouldn't require too much, if possible, we should make some contributions like writing docs or code.

In recent months, Trojan[2] vs V2ray + ws +tls + nginx(or caddy) is a pretty hot topic. If you don't know how to set it up. Google can help a lot. There's already many scripts on GitHub. These two methods are similar and pretty safe in our mind. If you can do some more research about this, which would make much more sense than v2ray + tcp only. Since many of us already got these by real usage years ago. 

Maybe you're new and haven't had much real experience, so when you said we were fucked after using DNN got some strong conclusions in the first place, most of us would argue about it. ML is awesome and used in many places in China already, so-called big data are mainly used for surveillance indeed, we have little good feeling about it if it's being used as like this. Of course everyone cares it. And if there's some conclutions, people will spread very quickly, trojan becomes hot now since we know v2ray only is not safe after last year's incidents, many vps are blocked since then. We got this by real experiences not by experiments, maybe we as normal users are not scientifc, but it works. 

[1] https://github.com/search?p=1&q=GFW&type=Repositories
[2] https://github.com/trojan-gfw/trojan

**Comment by ailinwang** [comment]
At the beginning I thought Mr. Su is from the authority and is here to guide technical bypassing into a swamp where censorship can be easily achieved. However, after reading his meaningless debate, I jump to the conclusion that he/she is such a crying baby that begs for a stop of research into the potential technologies that could very likely be used as a method for traffic sniffing.
The best defense is to attack, as Bruce Lee put it. Without knowing the mechanism behind which the sniffing works, one will **NEVER** know what vulnerabilities that V2RAY has. 
Please **BE SMART** ! @suhuaixing

**Comment by rickyzhang82** [comment]
I cherish Western Universal Values. Freedom of speech means everyone even those working for the evil regime is entitled to speak freely as well. In the past, I was furious to see so many fifty cents are active in Western social media like Facebook, Twitter, and Youtube to influence the view of people living beyond the wall. Now I'm not grumpy any more after I realized why Americans allow anyone, regardless of their motivation, to use their social platform. Because given free information access to anyone even fifty cents or hardcore commies, their value and ideology will change.

The debate as a rational and civilized human is an enlightenment process that benefits every participant. One thing I learned from @suhuaixing lengthy argument is that how cool kids in China today use V2Ray to bypass censorship. Now I change my course to research on TLS traffic.

I understood that English is the second language for all of us here. The adjective **meaningless** is a strong word. Even someone or I ask a **stupid** question or make an argument like a **fool**, it is great that someone else could point it out with solid evidence and rational reasoning.

I'm not pointing finger to say who is fifty cents or who is working for the evil regime. Nobody is an expert on everything. The more we discuss, the better informed we are.

**Comment by madeye** [comment]
Protocols like shadosocks and vmess share the same “flaw” that their packets are very random.

To identify those “random” packets, the easiest way is computing the entropy of the first packets of a TCP stream like this https://github.com/isofew/sssniff.

It’s a known issue for years. That’s why the DNN approach here is thought as “meaningless”, or at least “overkill”.

BTW, the right direction of future circumventing tools is obfuscating.

**Comment by nicholascw** [comment]
Throughout the thread I see nothing about any specific issue to the project but only discussion on your own research. I'll say this is the place to raise issues for V2Ray developers instead of a general discussion board, which you may found at [v2ray/discussion](https://github.com/v2ray/discussion/issues). Therefore I reckon this issue can be closed for now, if you think it's an issue for you that your packet can be detected and filtered out from normal traffic, than come up with a fix. I'll say that V2Ray itself, though widely used for firewall trespassing, it's still a general network utility that focus on efficiency, functionality and more, instead of just hiding your traffic. If that's what you previously thought about V2Ray you are in the wrong place.

Last thing I would like to say is that as long as it's a designed protocol, there should be a way to classify it out from others, it's just means and effort are uncertain. If you truly want to mix every sort of your packets together, just simply wrap it up with other protocols as one more layer out of it.
